Improve TS types

lishaduck opened the pull request:
Depends on #5
Here's the diff as GH doesn't play well with stacks from forks: <https://github.com/lishaduck/elm-solve-deps-wasm/compare/better-cargo+clippy...better-ts-types>

Removes `any` from the generated TS types.
The way the hashmap was handled made the minimal diff, but a more "correct" way of doing it would use `serde-wasm-bindgen` to serialize it.
The commit to update parameters should 100% make no difference to the generated wasm (as should the typo fix)
Dependency updates and the return type changes might make some differences, but I didn't notice tests. How should I check it works?

_**Closes:** #2_

P.S.: Correct me if I'm wrong, but I think the main consumer of this is elm-review. If elm-review does migrate to ESM, I'll be back! 😄

lishaduck committed:
chore: minimal dependency updates

Just what cargo bumps automatically.

lishaduck committed:
fix: typo in comment

lishaduck committed:
chore: better parameter types

lishaduck committed:
chore: better return type

lishaduck commented:
> The way the hashmap was handled made the minimal diff, but a more "correct" way of doing it would use `serde-wasm-bindgen` to serialize it.

I take this back. I accidentally said it was a function. Fixing that broke some other stuff, so I'll probably go back down this road.

lishaduck committed:
chore: remove outdated comment

lishaduck committed:
style: match pedantic style

Good ol' clippy found a thing or two.
This is a high quality codebase—these lints were only added a version or two ago.

lishaduck commented:
> > The way the hashmap was handled made the minimal diff, but a more "correct" way of doing it would use `serde-wasm-bindgen` to serialize it.
> 
> I take this back. I accidentally said it was a function. Fixing that broke some other stuff, so I'll probably go back down this road.

I take that back as well 🤣 I forgot to save, so the diagnostics didn't go away. This code compiles fine.

mpizenberg commented:
ahah you’re fast. So there are two main users of this, elm-review (the node package) and elm-test (the node package). So these are the two you should check with.

lishaduck commented:
> ahah you’re fast. So there are two main users of this, elm-review (the node package) and elm-test (the node package). So these are the two you should check with.

For tests, you mean?

mpizenberg commented:
For checking that nothing broke for them, and checking that they are fine with the type upgrade or if that messes things up with their setup.

lishaduck commented:
> For checking that nothing broke for them, and checking that they are fine with the type upgrade or if that messes things up with their setup.

👍 

I'm checking elm-review, as I made these changes by "copying" them over.
I'll check elm-test after that (might break tests if incorrect impls, but it's not typesafe).

mpizenberg commented:
For testing, most of the logic testing is done in the core library. This package is just a wrapper to provide a wasm package. If you can check that the two examples (offline and online) works, that would be enough I think. In addition to getting @jfmengels and @lydell on board with the changes.

lishaduck committed:
chore: add shebangs

For testing.

lishaduck committed:
chore: update lockfile format

NPM whines otherwise.

lishaduck committed:
fix: typos

lishaduck committed:
chore: remove extraneous catch bindings

lishaduck committed:
chore: update metadata

lishaduck commented:
> For testing, most of the logic testing is done in the core library. This package is just a wrapper to provide a wasm package. 

Makes sense.

> If you can check that the two examples (offline and online) works, that would be enough I think.

They work! (well, offline errors, but correctly.)

<details><summary>Logs</summary>
<p>

### Online

```console
Solution for test dependencies:
Offline solver failed, switching to online solver.
{
  direct: {
    'elm/core': '1.0.5',
    'elm/json': '1.1.3',
    'elm/random': '1.0.0',
    'elm/time': '1.0.0',
    'elm-explorations/test': '1.2.2',
    'ianmackenzie/elm-float-extra': '1.1.0',
    'ianmackenzie/elm-units': '2.10.0'
  },
  indirect: { 'elm/html': '1.0.0', 'elm/virtual-dom': '1.0.3' }
}
```

### Offline

```console
Directory "/Users/dukese01/.elm/0.19.1/packages/ianmackenzie/elm-float-extra does not exist
Not doing a request to the package server to find out existing versions. Please run at least once elm-test first.
Directory "/Users/dukese01/.elm/0.19.1/packages/ianmackenzie/elm-units does not exist
Not doing a request to the package server to find out existing versions. Please run at least once elm-test first.
Directory "/Users/dukese01/.elm/0.19.1/packages/ianmackenzie/elm-float-extra does not exist
Not doing a request to the package server to find out existing versions. Please run at least once elm-test first.

/Users/dukese01/Developer/elm-solve-deps-wasm/pkg/elm_solve_deps_wasm.js:252
            throw takeObject(r2);
            ^
Because there is no version of ianmackenzie/elm-float-extra in 1.1.0 <= v < 2.0.0 and ianmackenzie/elm-units-interval 2.3.0 depends on ianmackenzie/elm-float-extra 1.1.0 <= v < 2.0.0, ianmackenzie/elm-units-interval 2.3.0 is forbidden.
(Use `node --trace-uncaught ...` to show where the exception was thrown)

Node.js v20.12.2
```

</p>
</details> 

> In addition to getting @jfmengels and @lydell on board with the changes.

I can wait, elm-review's not blocked. 😁


P.S.: I mostly write Dart these days, and love that this uses PubGrub. Nice, familiar errors 😁

lishaduck committed:
chore: use node protocol

lishaduck committed:
chore: ts

Apparently Prettier and flow comments don't like each other.
Thus, it's a ts project now :)

lishaduck committed:
chore: esm-ify examples

lishaduck commented:
Presumably you'd like if I break up this PR?
I'm thinking
1. Update examples folder to a modern setup
2. Cargo changes
3. The changes to the .d.ts file generation

lishaduck committed:
style: prettier

lishaduck commented:
I've uploaded this to `github:lishaduck/elm-solve-deps-wasm-pkg` for testing, as npm has an longstanding bug with local path dependencies.

lishaduck reviewed (commented):
Ok, so all of `elm-review`'s test suite passes with these changes:  jfmengels/node-elm-review#195.

lydell commented:
node-test-runner unfortunately still uses Flow instead of TypeScript, so the change of types here does not affect it. Sounds good to improve the types, though!

lishaduck commented:
> node-test-runner unfortunately still uses Flow instead of TypeScript, so the change of types here does not affect it.
> Sounds good to improve the types, though!

Ok, thanks for the feedback. I still plan on testing though, as wasmbindgen has some newer features that allow these types, and so I've switched a few _very minor_ minor impl details.

lishaduck committed:
chore: bump msrv

Tidy up the code a bit, as there're no Rust consumers of this package.

lishaduck commented:
Simon said that the tests pass for elm-test-runner, so I think that's the last external blocker, we just need to resolve the remaining questions in #5.

mpizenberg commented:
I’m on a new machine (Mac M3) and was trying to recompile the project on the main branch first before checking out your changes. I’m not sure why, but I get this error while trying the online example:

<img width="865" alt="image" src="https://github.com/mpizenberg/elm-solve-deps-wasm/assets/2905865/53464091-37ff-4c5b-b5d6-a39739999694">

I have already compiled with wasm-pack the rust lib, and the `pkg/` directory is present at the correct location relative the the `package.json` file in the example folder. I’m not familiar enough with the JS ecosystem to know what this error is really about. Any idea?

mpizenberg commented:
AH, I had just forgotten to `npm install` 🤦
